crt/debugmalloc.c

```diff
@@ -0,0 +1,538 @@
+    return _calculate_block_size(header->alignment, header->size);
+}
+
+/* Doubly-linked list of headers */
+typedef struct _list
+{
+    header_t* head;
+    header_t* tail;
+} list_t;
+
+static list_t _list = {NULL, NULL};
+static myst_spinlock_t _spin = MYST_SPINLOCK_INITIALIZER;
+
+static void _list_insert(list_t* list, header_t* header)
+{
+    myst_spin_lock(&_spin);
+    {
+        if (list->head)
+        {
+            header->prev = NULL;
+            header->next = list->head;
+            list->head->prev = header;
+            list->head = header;
+        }
+        else
+        {
+            header->prev = NULL;
+            header->next = NULL;
+            list->head = header;
+            list->tail = header;
+        }
+    }
+    myst_spin_unlock(&_spin);
+}
+
+static void _list_remove(list_t* list, header_t* header)
+{
+    myst_spin_lock(&_spin);
+    {
+        if (header->next)
+            header->next->prev = header->prev;
+
+        if (header->prev)
+            header->prev->next = header->next;
+
+        if (header == list->head)
+            list->head = header->next;
+        else if (header == list->tail)
+            list->tail = header->prev;
+    }
+    myst_spin_unlock(&_spin);
+}
+
+MYST_INLINE bool _check_multiply_overflow(size_t x, size_t y)
+{
+    if (x == 0 || y == 0)
+        return false;
+
+    size_t product = x * y;
+
+    if (x == product / y)
+        return false;
+
+    return true;
+}
+
+static void _malloc_dump(size_t size, void* addrs[], int num_addrs)
+{
+    fprintf(stderr,"=== %lu bytes\n", size);
+    _dump_backtrace(addrs, num_addrs);
+}
+
+static void _dump(void)
+{
+    list_t* list = &_list;
+    bool found = false;
+    size_t blocks = 0;
+    size_t bytes = 0;
+
+    /* Count bytes allocated and blocks still in use */
+    for (header_t* p = list->head; p; p = p->next)
+    {
+        found = true;
+        blocks++;
+        bytes += p->size;
+    }
+
+    /* if any non-supressed blocks were found */
+    if (found)
+    {
+        fprintf(stderr,
+            "=== blocks in use: %zu bytes in %zu blocks\n", bytes, blocks);
+
+        for (header_t* p = list->head; p; p = p->next)
+            _malloc_dump(p->size, p->addrs, (int)p->num_addrs);
+
+        fprintf(stderr,"\n");
+    }
+}
+
+/*
+**==============================================================================
+**
+** Public definitions:
+**
+**==============================================================================
+*/
+
+void* malloc(size_t size)
+{
+    if (__crt_enable_debug_malloc)
+    {
+        void* block;
+        const size_t block_size = _calculate_block_size(0, size);
+
+        if (!(block = libc_malloc(block_size)))
+            return NULL;
+
+            /* fill block with 0xaa (allocated) bytes */
+#ifdef ENABLE_MALLOC_MEMSET
+        memset(block, 0xAA, block_size);
+#endif
+
+        header_t* header = (header_t*)block;
+        INIT_BLOCK(header, 0, size);
+        _check_block(header);
```

Comment:
Can this check ever fail?

Reply:
Yes, and when it does it prints an error to standard error and aborts.